In [97]:
import sys
import os

In [98]:
project_dir = os.path.dirname(os.getcwd())

In [99]:
sys.path.append(os.path.abspath(os.path.join(os.path.dirname("classifier.ipynb"), "..")))

## Import the Classes

In [149]:
from models.decisiontreeclassifier import DecisionTreeClassifier
from models.randomforestclassifier import RandomForestClassifier
from models.xgboostclassifier import XGBoostClassifier
from models.logisticregression import LogisticRegression

In [61]:
import pandas as pd
import numpy as np

## Load the data

In [62]:
heart = pd.read_csv(os.path.join(project_dir, "data/heart.csv"))

In [63]:
X = heart.drop(columns="target")
y = heart["target"]

In [64]:
n = len(X)

In [65]:
indices = np.arange(n)
np.random.shuffle(indices)

In [ ]:
s1 = int(0.75 * n)

In [89]:
X_train = X.iloc[indices[:s1],]
y_train = y.iloc[indices[:s1],]
X_test = X.iloc[indices[s1:],]
y_test = y.iloc[indices[s1:],]

## Logistic Regression

In [90]:
lg = LogisticRegression()

In [91]:
lg.fit(X_train, y_train)

/Users/vallurileelasaikrishna/Documents/from-scratch/models/logisticregression.py:16: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-z))


In [104]:
lg.accuracy(X_test, y_test)

/Users/vallurileelasaikrishna/Documents/from-scratch/.venv/lib/python3.13/site-packages/pandas/core/arraylike.py:402: RuntimeWarning: overflow encountered in exp
  result = getattr(ufunc, method)(*inputs, **kwargs)


np.float64(0.5758754863813229)

## Decision Tree

In [144]:
dtc = DecisionTreeClassifier()

In [147]:
dtc.fit(X_train, y_train)

In [148]:
dtc.accuracy_test(X_test, y_test)

np.float64(0.8871595330739299)

## Random Forest

In [159]:
rfc = RandomForestClassifier()

In [161]:
rfc.fit(X_train, y_train)

In [162]:
rfc.accuaracy_test(X_test, y_test)

np.float64(0.9260700389105059)

## XGBoost

In [163]:
xgbc = XGBoostClassifier()

xgb takes train and val at the same time for fit

In [ ]:
s1 = int(0.7 * n)
s2 = int(0.85 * n)

In [165]:
X_train = X.iloc[indices[:s1],]
y_train = y.iloc[indices[:s1],]
X_val = X.iloc[indices[s1:s2],]
y_val = y.iloc[indices[s1:s2],]
X_test = X.iloc[indices[s2:],]
y_test = y.iloc[indices[s2:],]

In [166]:
xgbc.fit(X_train, y_train, X_val, y_val)

Early stopping at tree 584


In [167]:
xgbc.accuracy(X_test, y_test)

np.float64(0.9514563106796117)

### Model Performance Comparison

| Model | Accuracy |
|--------|----------:|
| Logistic Regression | **0.5759** (57.59%) |
| Decision Tree | **0.8872** (88.72%) |
| Random Forest | **0.9261** (92.61%) |
| XGBoost | **0.9515** (95.15%) |